onnxruntime/test/python/onnxruntime_test_python.py

```diff
@@ -43,6 +43,126 @@ def testSetProviders(self):
             # confirm only CPU Provider is registered now.
             self.assertEqual(['CPUExecutionProvider'], sess.get_providers())
```

Comment:
I have concern on this test.

Currently, one ORT process can only use one CUDA device because of the extensive use of thread local variables in CUDA EP, but two processes can use two different devices. Until the limitation get removed, we can't really run device id tests like this. Even it can pass this time, it could be fragile.  I would prefer to have something easy to understand and explain.

Reply:
Thanks for bringing out this issue.

Just to make sure, you mentioned that "one ORT process can only use one CUDA device",
which means if I run following Python codes:
sess.set_providers(['CUDAExecutionProvider'], [{'device_id': 0}])
.....
sess.set_providers(['CUDAExecutionProvider'], [{'device_id': 1}])
.....

That will cause potential problems?
If the answer is yes, even though we confirm set_providers() can correctly set value, when user tires to call set_providers() to set different CUDA device ID within the same ORT process, it can still be possible to fail, right?

Reply:
Thanks for bringing out this issue.

Just to make sure, you mentioned that "one ORT process can only use one CUDA device",
which means if I run following Python codes:
sess.set_providers(['CUDAExecutionProvider'], [{'device_id': 0}])
.....
sess.set_providers(['CUDAExecutionProvider'], [{'device_id': 1}])
.....

That will cause potential problems?

If the answer is yes, even though we confirm set_providers() can correctly set value, when user tires to call set_providers() to set different CUDA device ID within the same ORT process, it can still be possible to fail, right?

Reply:
Yes.  You may check with Hector, I had a discussion with him months ago.

Reply:
I got a problem here.
Current python test case only calls sess.set_providers() which does CUDA EP initialization and not do the inference. 
Is CUDA EP initialization on different devices having potential issues as well?

Right now, I can't think of other ways to test this feature. 
Also, if our current ORT has this potential issue, can we still release this python API: sess.set_providers(['CUDAExecutionProvider'], [{'device_id': 0}])? since user could still set different devices within the same ORT process. 

@HectorSVC , do you have suggestions on this? Thank you.